lib/db/leveldb_transactions.go

```diff
@@ -85,107 +86,89 @@ func (t readWriteTransaction) insertFile(fk, folder, device []byte, file protoco
 // If the file does not have an entry in the global list, it is created.
 func (t readWriteTransaction) updateGlobal(gk, folder, device []byte, file protocol.FileInfo, meta *metadataTracker) bool {
 	l.Debugf("update global; folder=%q device=%v file=%q version=%v invalid=%v", folder, protocol.DeviceIDFromBytes(device), file.Name, file.Version, file.Invalid)
-	name := []byte(file.Name)
-	svl, _ := t.Get(gk, nil) // skip error, we check len(svl) != 0 later
 
 	var fl VersionList
-	var oldFile protocol.FileInfo
-	var hasOldFile bool
-	// Remove the device from the current version list
-	if len(svl) != 0 {
-		fl.Unmarshal(svl) // skip error, range handles success case
-		for i := range fl.Versions {
-			if bytes.Equal(fl.Versions[i].Device, device) {
-				if fl.Versions[i].Version.Equal(file.Version) && fl.Versions[i].Invalid == file.Invalid {
-					// No need to do anything
-					return false
-				}
+	if svl, err := t.Get(gk, nil); err == nil {
+		fl.Unmarshal(svl) // Ignore error, continue with empty fl
+	}
+	fl, removedFV, removedAt, insertedAt := fl.update(folder, device, file, t.db)
+	if insertedAt == -1 {
+		l.Debugln("update global; same version, global unchanged")
+		return false
+	}
 
-				if i == 0 {
-					// Keep the current newest file around so we can subtract it from
-					// the metadata if we replace it.
-					oldFile, hasOldFile = t.getFile(folder, fl.Versions[0].Device, name)
-				}
+	if insertedAt == 0 {
```

Comment:
One thing to take into account here that I had to fix up in `updateGlobal` for receive only, and that I *think* can be an issue already without the receive only stuff...

If we update a local file to a *lower* version than what is currently in there, what is the head of the version list may change without the insert being at index zero. For receive only folders this happens when the user does a "revert", but I think it could happen already today when a file gets unignored as we then tweak the version. Whether the new version becomes the head of the list or not depends on conflict resolution, and if it does *not* win the resolution it'll be inserted somewhere in the middle.

So, basically, if the version list is `{F1, F2, F3}` where F1 is our local file and the need is nil (we have the latest version), an update after unignore (or the upcoming revert operation) can make it `{F2, F1', F3}` or `{F2, F3, F1'}`, where `F1'` is the version-tweaked unignored file.

We need to make F2 the needed version then, despite the insert happening somewhere further down the list.

After receive only folders it becomes worse, although that's not your problem at the moment. :) The head of the list can then be a version we need, which gets withdrawn due to a change happening on that receive only device, and potentially another version from another device then becomes promoted to the head of the list and becomes the *new* needed file.

Reply:
Definitely true, `removedAt == 0` needs to be handled as well - thanks for catching that.
Doesn't that mean that global meta counting is broken already in this scenario?

Reply:
Yeah, I think so. I have it fixed but it's on top of a couple of other things, and of course it interferes with this PR.